### PR TITLE
pre-commit hooks for code style and lint check

### PR DESCRIPTION
**Pull request by beliaev-maksim**
## Changelog

* Added `pre-commit` configuration
* Ran `pre-commit` on all files
* Replaced `tox` calls to pre-commit
* Updated CI/CD pipeline for autorun on PRs

### CODE REVIEW

**Comment by rwcarlsen** [comment]
A couple initial thoughts.  I'm uncomfortable with having our CI checks not be trivial to run locally.  Before, if CI linting failed, I could just run `tox fmt` or `tox lint` to check and deal with things locally.  I'd prefer we keep that synchronized.  If these hooks in other repos change under our nose, things could break and I don't want to have to go digging to figure out why.  A second issue is the _vendor directory needs to remain untouched - those are files/libs we've copied from other projects and we don't want to modify them.

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
idea of pre-commit is that formatting is done during commit, that excludes a possibility of a bad commit. 
But that is definitely doable, we can add a tox hook for those who don't want to run a check "per commit"

`_vendor` -> will revert and exclude

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
try to either `pre-commit install` and just commit a file, it will be autofixed

or just run `tox -e lint`, which will trigger `pre-commit`

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
I did not tell that it will do `auto commit`, it will not

it just runs hooks during a commit process,it is described here https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks

**Comment by rwcarlsen** [comment]
Okay - I see.  This is (now) just using a (locally installed) `pre-commit` tool that I can use/run (if wanted) - that seems much more reasonable.

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
correct

and for everybody who wants to get this check every time can run `pre-commit install` and it will add a hook, every time you commit, it will run these checks allowing you to identify issues sooner.

here is an example:
image

**Comment by rwcarlsen** [comment]
Is there a way to leave `from foo import x, y, z` imports and the quotes (" vs ') alone?  That causes a lot of thrashing that will mess up git blame  (although @PietroPasotti has done a good deal of damage here with type checking :P).

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
here is a very good explanation why separated imports are better: https://github.com/asottile/reorder_python_imports#what-does-it-do
while I would stick to `isort` since it is PyCQA repo


regarding quotes, the same thing. This is a default behaviour of `black` owned by psf (python software foundation). 
Sticking to the community standard is always better (like you eyes get used to the specific format :) and you can jump from every open source project to another without a need to relearn new styles, idea is like `go fmt` )

**Comment by rwcarlsen** [comment]
I understand.  Part of the power of gofmt is that it has been the standard since the very beginning of go.  Python has thousands of projects with thousands of formats.  I'm reluctant to change the project style just to suit a (partial) community norm if we already have a way of auto-formatting for folks who aren't accustomed to any differences.  In my view the cost-benefit for us is dubious here.

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
`black` is owned by psf, so, I would say this one is the closed to be named official.
I don't think this transition will be painful. In return we will get official python community supported formatter.

**Comment by jnsgruk** [comment]
We went through this a few months ago and settled on autopep8.

Happy for precommit, but let's revert the formatting changes from black and keep that for a separate PR. 

I'd be keen to consolidate, too. We already have tool config in pyproject.toml, I'd rather not have it separately in precommit - that's just begging to get out of sync....

**Comment by pengale** [comment]
I was out on Monday, and missed the initial discussion ... 

I think that this is two PRs.

1) Add pre-commit hooks. I am +1 to this -- it is something that we have talked about implementing, and I appreciate the contribution!

2) Moving from autopep8 to black. I am -1 to this. The project has already taken on a fair amount of risk with the typing changes, and I do not think that it is a good idea to take on the risk of additional sweeping formatting style changes. On top of this, the decision to go with `autopep8` was made based on feedback from multiple stakeholders -- it wouldn't be fair to reverse that decision without buy in across the ops community.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    hooks:
+      - id: blacken-docs
+        additional_dependencies: [black]
+
+  - repo: https://github.com/pycqa/isort
+    rev: 5.10.1
+    hooks:
+      - id: isort
+        name: isort (python)
+        args: ["--force-single-line-imports", "--profile", "black"]
```
Comment (jnsgruk) [inline .pre-commit-config.yaml:25]: I would drop this and revert the changes it's made (feels a bit noisy). There is already some basic `isort` config in the `pyproject.toml`

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    rev: v1.7.0
+    hooks:
+      - id: autopep8
+
+  - repo: https://github.com/asottile/blacken-docs
+    rev: v1.12.1
+    hooks:
+      - id: blacken-docs
+        additional_dependencies: [black]
+
```
Comment (jnsgruk) [inline .pre-commit-config.yaml:19]:
```suggestion
```

**Comment by beliaev-maksim** [comment]
@rwcarlsen @pengale @jnsgruk 
I talked with Jon yesterday, for the moment we revert to `autopep8`

issues that I still see, 
I fully understand your willing to use only `pyproject.toml`, but I am worried that official `flake8` does not support it and we have to use badly maintained `pyproject-flake8`. For example, it does not support latest `flake8` and this is untouched for months.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (beliaev-maksim) [reply .pre-commit-config.yaml:25]: @jnsgruk 
see https://github.com/canonical/operator/pull/825#issuecomment-1238284247

but I can move `isort` config to pyproject.toml

**Review by jnsgruk — CHANGES_REQUESTED** [review]
I think I'm -1 on this. At the moment, all of the charm teams (and thus the people who work on/consume this lib) are using the setup we have in place with `tox`. While I can see some benefit to having these checks run in a pre-commit hook, I think we can find a lighter way of doing it that doesn't introduce *another* dev tool.

**Comment by beliaev-maksim** [comment]
@jnsgruk 
and we still use `tox`, but `tox` calls `pre-commit` under the hood

**Comment by beliaev-maksim** [comment]
> pre-commit linting/formatting hooks would be absolutely welcome. But I don't think we are ready to change formatting tools or code style at this time.


and the latest commits address it :)
I used flake, autopep8, isort 
there are two minor addition, blacken-docs, which touches only code blocks in docs. I believe that everybody will agree that this is not harmful. And autoflake, that checks for unused imports, also very minor

**Comment by rwcarlsen** [comment]
The PR still includes all the import line changes for which necessary discussion/agreement has not occurred - let's walk those back.  Notably, we're not wholesale against formatting adjustments, but such changes would need to be appropriately considered in a separate issue where we can discuss with and loop in interested parties.

**Comment by beliaev-maksim** [comment]
@rwcarlsen 
agree, let's discuss the topic of `black` and `isort` in a wider group later

**Comment by jnsgruk** [comment]
Closing for now, will revisit when team has more bandwidth.
